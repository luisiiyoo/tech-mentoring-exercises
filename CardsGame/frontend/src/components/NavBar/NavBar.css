.NavBar-Icon {
  font-size: 2em;
  vertical-align: middle;
}

.sidenav---sidenav---_2tBP {
  background: var(--sidenav-background-app);
}
