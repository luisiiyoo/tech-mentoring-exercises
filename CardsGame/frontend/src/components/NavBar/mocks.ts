import { NavItem } from '../../model/navItem';

export const navItemMock: NavItem[] = [
  {
    route: 'my-home',
    title: 'My Home',
    iconClass: 'fa fa-fw fa-heartbeat',
  },
  {
    route: 'free-code-camp',
    title: 'Free Code Camp',
    iconClass: 'fa fa-fw fa-beer',
    childs: [
      {
        idChild: '1',
        route: 'sub-page',
        title: 'SubPage',
      },
    ],
  },
];

export const stateMock = {
  main: {
    route: 'home',
    expand: true,
  },
};
