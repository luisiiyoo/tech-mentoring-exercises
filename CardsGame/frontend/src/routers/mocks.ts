import { NavItem } from '../model/navItem';

export const navBarItemsMock: NavItem[] = [
  {
    route: 'home',
    title: 'Home',
    iconClass: 'fa fa-fw fa-home',
  },
  {
    route: 'engineering',
    title: 'Engineering',
    iconClass: 'fa fa-fw fa-graduation-cap',
    childs: [
      {
        idChild: '1',
        route: 'basic-back-end',
        title: 'Basic Back-End',
      },
      {
        idChild: '2',
        route: 'avanced-back-end',
        title: 'Avanced Back-End',
      },
      {
        idChild: '3',
        route: 'basic-front-end',
        title: 'Basic Front-End',
      },
      {
        idChild: '4',
        route: 'avanced-front-end',
        title: 'Avanced Front-End',
      },
      {
        idChild: '5',
        route: 'qa',
        title: 'Q&A',
      },
    ],
  },
];

export const stateMock = {
  main: {
    expand: true,
  },
  topicDetails: {
    visible: false,
    topic: {
      id: -1,
      name: '',
      level: -1,
      shortDescription: '',
      description: '',
      type: '',
      tags: [],
      metadata: {},
      resources: [
        {
          id: -1,
          name: '',
          shortDescription: '',
          description: '',
          url: '',
          complexity: '',
          tags: [],
          metadata: {},
        },
      ],
    },
  },
};
