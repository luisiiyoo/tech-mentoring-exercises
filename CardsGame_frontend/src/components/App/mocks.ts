export const navItemsMock = [
  {
    route: 'my-home',
    title: 'My Home',
    iconClass: 'fa fa-fw fa-heartbeat',
  },
  {
    route: 'free-code-camp',
    title: 'Free Code Camp',
    iconClass: 'fa fa-fw fa-beer',
    childs: [
      {
        idChild: '1',
        route: 'sub-page',
        title: 'SubPage',
      },
    ],
  },
];
