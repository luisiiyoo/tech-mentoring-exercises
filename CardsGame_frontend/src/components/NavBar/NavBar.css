.NavBar-Icon {
  font-size: 2em;
  vertical-align: middle;
}
