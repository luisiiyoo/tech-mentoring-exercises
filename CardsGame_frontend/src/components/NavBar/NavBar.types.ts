import { NavItem } from 'src/model/navItem';

interface History {
  push(path: string): void;
  location: {
    pathname: string;
  };
}
export interface OnChangeRoute {
  (selectedRoute: string): void;
}

export interface OnSelectNavItem {
  (selected: string, history: History): void;
}

export interface NavBarProps {
  navBarTitle: string;
  history: History;
  navBarItems: NavItem[];
}

export interface RenderNavItem {
  (item: NavItem, isChild?: boolean): void;
}
