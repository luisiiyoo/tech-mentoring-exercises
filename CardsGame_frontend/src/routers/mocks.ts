import { NavItem } from '../model/navItem';

export const navBarItemsMock: NavItem[] = [
  {
    route: 'home',
    title: 'Home',
    iconClass: 'fa fa-fw fa-home',
  },
  {
    route: 'engineering',
    title: 'Engineering',
    iconClass: 'fa fa-fw fa-graduation-cap',
    childs: [
      {
        journeyID: '1',
        route: 'basic-back-end',
        title: 'Basic Back-End',
      },
      {
        journeyID: '2',
        route: 'avanced-back-end',
        title: 'Avanced Back-End',
      },
      {
        journeyID: '3',
        route: 'basic-front-end',
        title: 'Basic Front-End',
      },
      {
        journeyID: '4',
        route: 'avanced-front-end',
        title: 'Avanced Front-End',
      },
      {
        journeyID: '5',
        route: 'qa',
        title: 'Q&A',
      },
    ],
  },
];

export const stateMock = {
  main: {
    expand: true,
  },
  topicDetails: {
    visible: false,
    topic: {
      id: -1,
      name: '',
      level: -1,
      shortDescription: '',
      description: '',
      type: '',
      tags: [],
      metadata: {},
      resources: [
        {
          id: -1,
          name: '',
          shortDescription: '',
          description: '',
          url: '',
          complexity: '',
          tags: [],
          metadata: {},
        },
      ],
    },
  },
};
