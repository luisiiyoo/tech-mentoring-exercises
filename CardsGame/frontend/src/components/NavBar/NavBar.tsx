import React from 'react';
import { useSelector, useDispatch } from 'react-redux';
import SideNav from '@trendmicro/react-sidenav';
import {
  setSelectedRoute,
  setExpandNavBar,
} from 'src/redux/actions/MainComponent';
import { NavItem } from 'src/model/navItem';
import NavBarHeader from 'src/components/NavBarHeader';
import { NavBarProps, OnSelectNavItem, RenderNavItem } from './NavBar.types';
import '@trendmicro/react-sidenav/dist/react-sidenav.css';
import './NavBar.css';

const styleNavItemText = { fontSize: '1.3em', verticalAlign: 'middle' };
const styleNavSubItemText = {
  paddingLeft: 10,
  fontSize: '1.0em',
  verticalAlign: 'middle',
};

const onSelectNavItem: OnSelectNavItem = (selected, history) => {
  const {
    location: { pathname },
  } = history;
  const to = '/' + selected;
  if (pathname !== to) history.push(to);
};

const renderNavItem: RenderNavItem = (item: NavItem, isChild = false) => {
  const completeRoute = item.route;
  const childs = item.childs || [];
  return (
    <SideNav.NavItem
      eventKey={completeRoute}
      key={completeRoute}
      data-testid={completeRoute}
    >
      {item.iconClass && (
        <SideNav.NavIcon>
          <i className={`${item.iconClass} NavBar-Icon`} />
        </SideNav.NavIcon>
      )}
      <SideNav.NavText style={isChild ? styleNavSubItemText : styleNavItemText}>
        {item.title}
      </SideNav.NavText>
      {childs.map((child) => renderNavItem(child, true))}
    </SideNav.NavItem>
  );
};

const NavBar: React.FC<NavBarProps> = ({
  history,
  navBarItems,
  navBarTitle,
}) => {
  const { route, expand } = useSelector((state) => state.main);
  const dispatch = useDispatch();

  return (
    <SideNav
      expanded={expand}
      onSelect={(selected) => {
        onSelectNavItem(selected, history);
        dispatch(setSelectedRoute(selected));
      }}
      onToggle={(expanded) => {
        dispatch(setExpandNavBar(expanded));
      }}
    >
      <SideNav.Toggle data-testid="toggleNavBar" />
      <NavBarHeader expanded={expand} title={navBarTitle} />
      <SideNav.Nav selected={route}>
        {navBarItems.map((item) => renderNavItem(item))}
      </SideNav.Nav>
    </SideNav>
  );
};

export default NavBar;
