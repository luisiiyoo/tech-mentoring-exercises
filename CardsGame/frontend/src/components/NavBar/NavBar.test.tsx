import React from 'react';
import { cleanup, render, fireEvent } from '@testing-library/react';
import { useSelector } from 'react-redux';
import NavBar from './NavBar';
import { stateMock, navItemMock } from './mocks';

let historyMock;

jest.mock('react-redux', () => ({
  useSelector: jest.fn(),
  useDispatch: () => jest.fn(),
}));

describe('NavBar', () => {
  beforeEach(() => {
    historyMock = { push: jest.fn(), location: { pathname: '/' } };
    useSelector.mockImplementation((callback) => callback(stateMock));
  });

  afterEach(() => {
    cleanup();
    useSelector.mockClear();
  });

  it('should take a snapshot', () => {
    const { asFragment } = render(
      <NavBar
        history={historyMock}
        navBarItems={navItemMock}
        navBarTitle={`MyApp`}
      />,
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('should navIcons update the navigation history', () => {
    const { getByTestId } = render(
      <NavBar
        history={historyMock}
        navBarItems={navItemMock}
        navBarTitle={`MyApp`}
      />,
    );
    // This test only contemplates two levels of navItems
    navItemMock.forEach(({ childs, route }) => {
      if (!!childs && childs.length > 0) {
        childs.forEach((child) => {
          const idChild = child.route;
          fireEvent.click(getByTestId(idChild));
          expect(historyMock.push).toHaveBeenLastCalledWith(`/${idChild}`);
        });
      } else {
        fireEvent.click(getByTestId(route));
        expect(historyMock.push).toHaveBeenLastCalledWith(`/${route}`);
      }
    });
  });
});
