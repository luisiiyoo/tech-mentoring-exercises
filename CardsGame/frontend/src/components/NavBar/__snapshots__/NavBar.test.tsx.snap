// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NavBar should take a snapshot 1`] = `
<DocumentFragment>
  <nav
    class="sidenav---sidenav---_2tBP sidenav---expanded---1KdUL"
  >
    <button
      aria-expanded="true"
      class="sidenav---sidenav-toggle---1KRjR"
      data-testid="toggleNavBar"
      role="button"
    >
      <span
        class="sidenav---icon-bar---u1f02"
      />
      <span
        class="sidenav---icon-bar---u1f02"
      />
      <span
        class="sidenav---icon-bar---u1f02"
      />
    </button>
    <div
      class="NavBarHeader"
      data-testid="headerContainer"
      style="display: block;"
    >
      <div
        class="NavBarHeader-Title"
        data-testid="headerTitle"
      >
        MyApp
      </div>
    </div>
    <div
      class="sidenav---sidenav-nav---3tvij sidenav---expanded---1KdUL"
      role="menu"
    >
      <div
        class="sidenav---sidenav-navitem---uwIJ-"
        role="presentation"
      >
        <div
          class="sidenav---navitem---9uL5T"
          data-testid="my-home"
          role="menuitem"
          tabindex="-1"
        >
          <div
            class="sidenav---navicon---3gCRo"
          >
            <i
              class="fa fa-fw fa-heartbeat NavBar-Icon"
            />
          </div>
          <div
            class="sidenav---navtext---1AE_f"
            style="font-size: 1.3em; vertical-align: middle;"
          >
            My Home
          </div>
        </div>
      </div>
      <div
        class="sidenav---sidenav-navitem---uwIJ- sidenav---expandable---3_dr7"
        role="presentation"
      >
        <div
          class="sidenav---navitem---9uL5T"
          data-testid="free-code-camp"
          role="menuitem"
          tabindex="-1"
        >
          <div
            class="sidenav---navicon---3gCRo"
          >
            <i
              class="fa fa-fw fa-beer NavBar-Icon"
            />
          </div>
          <div
            class="sidenav---navtext---1AE_f"
            style="font-size: 1.3em; vertical-align: middle;"
          >
            Free Code Camp
          </div>
        </div>
        <div
          class="sidenav---sidenav-subnav---1EN61"
          role="menu"
        >
          <div
            class="sidenav---sidenav-subnavitem---1cD47"
            role="heading"
          >
            Free Code Camp
          </div>
          <div
            class="sidenav---sidenav-subnavitem---1cD47"
            role="presentation"
          >
            <div
              class="sidenav---navitem---9uL5T"
              data-testid="sub-page"
              role="menuitem"
              tabindex="-1"
            >
              <div
                class="sidenav---navtext---1AE_f"
                style="padding-left: 10px; font-size: 1.0em; vertical-align: middle;"
              >
                SubPage
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </nav>
</DocumentFragment>
`;
